### PR TITLE
Fix SLT parser

### PR DESCRIPTION
**Pull request by jamii**
This is working towards fixing stuff I broke in https://github.com/MaterializeInc/materialize/pull/2212/, but also seems to have fixed some pre-existing parser bugs that were hiding legit test failures eg 

```
query I nosort
SELECT 1 FROM t1 WHERE '1' IN ( NULL, 1 )
----
1
1
1
```

```
jamie=> CREATE TABLE t1( x INTEGER, y TEXT );
CREATE TABLE
jamie=> INSERT INTO t1 VALUES(1,'true');
INSERT 0 1
jamie=> INSERT INTO t1 VALUES(0,'false');
INSERT 0 1
jamie=> INSERT INTO t1 VALUES(NULL,'NULL');
INSERT 0 1
jamie=> SELECT 1 FROM t1 WHERE 1 IN ( NULL, 1 );
 ?column? 
----------
(0 rows)
```

### CODE REVIEW

**Comment by jamii** [comment]
Current slt-fast results: `FAIL: parse-failure=1 plan-failure=4 output-failure=59 bail=1 success=49802 total=49867`

Before parser breakage: `success=49873 total=49873`

We're still parsing less tests total, and some of them are also legit broken but were passing before? Mysterious.

**Comment by jamii** [comment]
Ok, the above worked as of 04ae49cbd67be9ee61499c5f5d909957d5523eb7 so this is a very new bug.

**Comment by jamii** [comment]
Now at `FAIL: plan-failure=3 output-failure=59 success=49812 total=49874` so I think the parser is ok and we just have some planner regressions from the last day or two.

**Comment by jamii** [comment]
@benesch I'm not sure how best to merge this. Should we fix the errors first?

File: test/sqllogictest/cockroach/union.slt [hunk 1]
```diff
@@ -257,7 +257,7 @@ CREATE TABLE b (a INTEGER PRIMARY KEY)
 
 query I
 SELECT * FROM a UNION ALL SELECT * FROM b
-
+----
```
Comment (benesch) [inline test/sqllogictest/cockroach/union.slt]: Are there not dozens more examples of this in the cockroach files? I seem to recall it being a fairly common pattern, and so one we might need to support.

Comment (jamii) [reply test/sqllogictest/cockroach/union.slt]: I just checked and there are no others. Stuff with `query error` often doesn't have `----` but that's a different path in the parser.

Comment (benesch) [reply test/sqllogictest/cockroach/union.slt]: 👍

**Comment by benesch** [comment]
Pushed up a commit to fix at least one of the plan failures by fixing our unary +- precedence. Looking into the output failures now.

**Comment by benesch** [comment]
aaaaargh

```
FAIL: output-failure=1 success=49881 total=49882
```

**Comment by jamii** [comment]
@benesch That last test might be wrong. It was recording an incompatibility with postgres that was assumed to be down to precision choices, but might also have had some precedence issues.

**Comment by benesch** [comment]
Oh I see. Then my latest fix is wrong. 1s.

**Comment by benesch** [comment]
Ok, adjusted. Hopefully this passes CI. 🤞

**Pull request merged**